src/verifiers/w3c-did/openAttestationW3CDIDProof.ts

```diff
@@ -0,0 +1,82 @@
+import * as ethers from "ethers";
+import { v2, WrappedDocumentWithProof } from "@govtechsg/open-attestation";
+import { resolveDID } from "./resolveDID";
+import { Verifier } from "../../types/core";
+import { Authentication } from "../../types/w3c-did";
+import { SUPPORTED_DID_AUTH, ETHR_DID_METHOD, SUPPORTED_PROOF_TYPES } from "../../config";
+import { OpenAttestationDocumentSignedCode } from "../../types/error";
+
+const name = "openAttestationW3CDIDProof";
+const type = "DOCUMENT_STATUS";
+
+export const openAttestationW3CDIDProof: Verifier<WrappedDocumentWithProof<v2.OpenAttestationDocument>> = {
+  skip: () => {
+    return Promise.resolve({
+      status: "SKIPPED",
+      type,
+      name,
+      reason: {
+        code: OpenAttestationDocumentSignedCode.SKIPPED,
+        codeString: OpenAttestationDocumentSignedCode[OpenAttestationDocumentSignedCode.SKIPPED],
+        message: `Document does not have a proof block`
+      }
+    });
+  },
+  test: document => {
+    return Object.keys(document).includes("proof");
+  },
+  verify: async document => {
+    try {
+      const { proof, signature } = document;
+      const didDoc = await resolveDID(proof.verificationMethod);
+
+      // Confirm this type of proof is currently supported
+      if (!SUPPORTED_PROOF_TYPES.includes(proof.type)) throw new Error(`Proof type: ${proof.type} is not supported.`);
+
+      // If the didDoc has an auth method we currently support
+      const supportedAuth = didDoc?.authentication?.filter((auth: Authentication) =>
+        SUPPORTED_DID_AUTH.includes(auth.type)
+      );
+      if (!supportedAuth?.length) throw new Error("Issuer DID cannot be authenticated, no supported auth in didDoc.");
+
+      // Get the correct pub key from the ver method
+      const pubKey: any = didDoc.publicKey.find(key => key.id === proof.verificationMethod);
+      // The exact verification may vary based on did-method (did:<method>:<namespace uuid>)
+      const didMethod = proof.verificationMethod.split(":")[1];
+      let proofValid = false;
+      if (didMethod === ETHR_DID_METHOD) {
+        // Existing targetHash is being signed
+        const msg = signature.targetHash;
+        const { ethereumAddress } = pubKey;
+        const recoverAddress = ethers.utils.verifyMessage(msg, proof.signature);
+        proofValid = recoverAddress?.toLowerCase() === ethereumAddress?.toLowerCase();
```

Comment:
I would expect recoverAddress and ethereumAddress to not be optional here :thinking:

Reply:
Correct re `recoveredAddress`, updated.  For `ethereumAddress` it is optional on the `PublicKey` type because there are other types of keys that do not have Ethereum Addresses, ie. commonly ed25519 keys are used.  I've updated to check for presence of `ethereumAddress` first now once we know we are using the `ethr` did-method.